feat: gateway controller

shaneutt opened the pull request:
**What this PR does / why we need it**:

This adds a `Gateway` controller as the first stage of [Gateway APIs](https://github.com/kubernetes-sigs/gateway-api) support in the KIC.

**Which issue this PR fixes**

Resolves https://github.com/Kong/kubernetes-ingress-controller/issues/1945

**Special notes for your reviewer**:

This iteration doesn't provide any end-user functionality, and is simply the first iteration in [Gateway Milestone 1](https://github.com/Kong/kubernetes-ingress-controller/milestone/21).

As per the [KEP](https://github.com/Kong/kubernetes-ingress-controller/blob/main/keps/0002-kong-gateway-api.md) this iteration is limited to "pre-provisioned" mode, meaning an existing Kong Gateway must exist in order to "provision" `Gateway` resources.

**PR Readiness Checklist**:

- [x] the `CHANGELOG.md` release notes have been updated

shaneutt commented on internal/controllers/gateway/gateway_controller.go:
```diff
@@ -0,0 +1,597 @@
+		r.Log.Error(fmt.Errorf("unexpected object type in gatewayclass watch predicates"), "expected", "*gatewayv1alpha2.GatewayClass", "found", reflect.TypeOf(obj))
+		return false
+	}
+	if gwc.Spec.ControllerName == ControllerName {
+		return true
+	}
+	return false
```
```suggestion
	return gwc.Spec.ControllerName == ControllerName
```

mflendrich commented on internal/manager/setup.go:
```diff
@@ -42,7 +43,26 @@ func setupLoggers(c *Config) (logrus.FieldLogger, logr.Logger, error) {
+	additionalCacheNamespaces := make([]string, 0)
+
+	// gateway APIs controller options setup
+	if c.FeatureGates["Gateway"] {
```
This is exactly the kind of branching I was concerned about when I was pushing hard against feature gating :thinking:

I don't have an obvious idea for a fix; let's try to think about it tomorrow.

shaneutt replied:
Gateway is a special case because it needs to be aware of the `PublishService` at this stage in the implementation. I don't expect us to have to make any other special adjustments. I also don't think long term as the implementation grows we'll continue to have that limitation, or it least it doesn't need to continue working in the very strict way it does now.

I thought long and hard on this issue before finally implementing this here and in avoidance of building an extra and separate cache just for the Gateway controller (which comes with its own problems) this seemed like the lightest touch for something that is not likely to (and if nothing else I don't think required to) last long term.

shaneutt replied:
We're considering this one a blocker for merge after having discussed it in a zoom meeting today.

mflendrich commented on internal/manager/setup.go:
```diff
@@ -52,23 +72,23 @@ func setupControllerOptions(logger logr.Logger, c *Config, scheme *runtime.Schem
-	switch len(c.WatchNamespaces) {
-	case 0:
-		// watch all namespaces
+	if len(c.WatchNamespaces) == 0 {
```
Could we leave the logic as simple as it was, and use a concatenation of WatchNamespaces and the additional namespaces altogether as the input list of namespaces?

shaneutt replied:
No because it would change the behavior: we need to specifically know if `c.WatchNamespaces` was zero because that is our indication that we're going to watch ALL namespaces, which supports the Gateway caching needs anyway. If we did any appending or joining of the two for identification we would end up with a broken "all" mode.

mflendrich commented on internal/controllers/gateway/gateway_controller.go:
```diff
@@ -0,0 +1,594 @@
+	obj *gatewayv1alpha2.Gateway,
+) (bool, error) {
+	if !readyConditionExistsForObservedGeneration(obj) {
+		obj.Status.Listeners = convertListenerToListenerStatus(obj)
```
I'd be uncomfortable merging this PR without an integration test that validates that the status fields get populated appropriately.

shaneutt replied:
This test already existed to do that: https://github.com/Kong/kubernetes-ingress-controller/pull/1986/files#diff-1cde2520ec6baa65f40a27801dbee1c6d60045bfdc2dfa1c36aec9785ee04211R101

mflendrich reviewed (changes requested):
First round - this is a huge PR so I expect more to come.

rainest reviewed (commented):
Initial review is that between this being a completely new (and particularly unique) controller and my limited understanding of GWAPI details I probably want a walkthrough of this on a call to review this effectively. Following are some initial comments that are 80% bikeshedding (though I do think the "provisioned" one is important) and 20% primer questions for a walkthrough.

Do we need to get rid of nakedreturn linting? Personal preference but I don't find them particularly readable, and it seems like if anything we should be exempting the individual instances if we really do want it somewhere (though I don't think there's ever any case where it's difficult to explicitly return).

gateway_controller.go differs considerably from the other controllers, which we outline in `hack/generators/controllers/networking/main.go` and template to create `internal/controllers/configuration/zz_generated_controllers.go`. This looks probably necessary, though it'd probably help to cover how and why it differs, especially for stuff that exists elsewhere in the codebase in a different form (class filtering and publish service information). AFAIK GWAPI differences (a controller possibly managing multiple gateways rather than exposing everything on one Ingress) probably necessitate this, but I don't have a good enough knowledge of them to say for sure whether the implementation looks correct. I'm further unsure how, if at all, this would relate to us properly implementing Ingress v1 IngressClass handling.

Is "Pre-Provisioned" and "Provisioned" official nomenclature (the annotations are ours and I didn't see mention of it in a quick search of the GWAPI docs, so guessing no)? My reading of the code suggests that the latter means the Gateway is not actually, well, provisioned, but rather something that needs to be provisioned, whereas my reading of the actual words is that they mean the same thing. Not immediately sure what would be better offhand--maybe "pre-provisioned" versus "controller-provisioned" or "externally-managed" and "controller-managed" to clarify the actor? Would maybe be easier to say knowing roughly how the controller would handle a gateway that isn't already provisioned in the future.

More naming nits: `obj` makes me think of generic K8S objects (when you actually don't know the GVK) and confused me in the longer functions; I had to stop a few times and scroll back to the signature to realize that it was the Gateway being reconciled or whatever. `gatewayObject` or just `gateway`/`gatewayClass`/etc. as appropriate?

This seems to have parallel implementations of (or at least related to) existing watch namespace and publish service functionality. How/why do GWAPIs need to handle these differently?

rainest commented on internal/controllers/gateway/gateway_controller.go:
```diff
@@ -0,0 +1,594 @@
+		return err
+	}
+
+	// watch for updates to gatewayclasses, if any gateway classes change, enqueue
```
Do we eventually end up with a separate GatewayClass reconciler? I don't think we have anything else where a change in one resource necessitates an update in another, which seems necessary here--the others aren't like this, but because they don't have to be. Proxy Service and Status is a sort-of exception but it's handled elsewhere.

shaneutt replied:
> This is a significant departure from the model where one reconciler watches one kind

This is a common implementation strategy among our neighbor gateway implementations to ensure that updates on Gateways occur in response to GatewayClass changes.

> Is this intentional?

As per the comment explaining why this is being done here, yes this is intentional.

> Do we eventually end up with a separate GatewayClass reconciler?

In time we most likely end up with a separate GatewayClass reconciler as our implementation grows, but at least for the purposes of this iteration it's not yet needed to start integrating webhook validation and `HTTPRoute`.

shaneutt replied:
We discussed this further in our zoom today, let me know if there's any more follow up needed here.

shaneutt commented on .golangci.yaml:
```diff
@@ -24,7 +24,6 @@ linters:
   - ineffassign
   - megacheck
   - misspell
-  - nakedret
```
I don't think this linter adds enough value to be worth keeping in place. It prescribes that functions which use variable declarations in the return signature to be somehow "unsafe" and picks what I can only assume is a semi-arbitrary line length to say when they become unsafe. I would rather we turn this off and allow the language feature to be used as intended.

shaneutt replied:
Since it was brought up multiple times, I'm just conceding on this one it's not worth the time. Changed in 0fbe4db1ce844331daf762d908784a36c2463266

shaneutt commented on config/rbac/role.yaml:
```diff
@@ -202,6 +202,40 @@ rules:
+  - list
+  - patch
+  - update
+  - watch
```
It's not, just something left over from the defaults, changed in 9b4680ebd9cce991e3bd2902910464c636b83a92

shaneutt commented on internal/controllers/gateway/gateway_controller.go:
```diff
@@ -0,0 +1,594 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
```
removed in c23d1276940dcd3d52fa67e102747700a287ba77

shaneutt commented on internal/controllers/gateway/gateway_controller.go:
```diff
@@ -0,0 +1,594 @@
+	r.publishServiceRef = types.NamespacedName{
+		Namespace: publishServiceSplit[0],
+		Name:      publishServiceSplit[1],
+	}
```
done in 9b32be801f0c609e8f9f628701466587e167ffd2 and added tests

shaneutt commented on internal/controllers/gateway/gateway_controller.go:
```diff
@@ -0,0 +1,594 @@
+	r.allowedRoutes.Namespaces = &namespaces
+
+	// convert the service ref into a NamespacedName
+	publishServiceSplit := strings.Split(r.PublishService, "/")
```
done in 9b32be801f0c609e8f9f628701466587e167ffd2

shaneutt commented on internal/controllers/gateway/gateway_controller.go:
```diff
@@ -0,0 +1,594 @@
+		r.Log.Error(fmt.Errorf("unexpected object type in gateway watch predicates"), "expected", "*gatewayv1alpha2.Gateway", "found", reflect.TypeOf(obj))
+		return false
+	}
+	gwc := &gatewayv1alpha2.GatewayClass{}
+	if err := r.Client.Get(context.Background(), client.ObjectKey{Name: string(gw.Spec.GatewayClassName)}, gwc); err != nil {
+		r.Log.Error(err, "could not retrieve gatewayclass", "gatewayclass", gw.Spec.GatewayClassName)
+		return false
+	}
+	return gwc.Spec.ControllerName == ControllerName
```
The controller runtime client present here is a cached client, so while the general answer to your question is "yes" the additional context is we're already doing that.

shaneutt commented on internal/controllers/gateway/gateway_controller.go:
```diff
@@ -0,0 +1,594 @@
+		r.Log.Error(err, "failed to list gateways for gatewayclass in watch", "gatewayclass", gwc.GetName())
+		return
+	}
+	for _, gw := range gateways.Items {
+		if string(gw.Spec.GatewayClassName) == gwc.GetName() {
+			recs = append(recs, reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Namespace: gw.Namespace,
+					Name:      gw.Name,
+				},
+			})
+		}
+	}
```
There's only two instances of a pattern at all like this and they both are unique from eachother. Given that plus given the fact that the integration tests are explicitly testing these (by verifying gatewayclasses posted AFTER connected gateways are posted results in the gateways getting updated in one case, and verifying that gateways connected to the publish service are updated when the publish service is updated in the second case). I don't see a strong reason to tear these up.

shaneutt commented on internal/controllers/gateway/gateway_controller.go:
```diff
@@ -0,0 +1,594 @@
+// -----------------------------------------------------------------------------
+
+var (
+	// type pointers for helper functions
+	gatewayIPAddrType   = gatewayv1alpha2.IPAddressType
+	gatewayHostAddrType = gatewayv1alpha2.HostnameAddressType
```
The answer to both questions is because the source is a `const` and we need to be able to take pointers to the value.

shaneutt commented on internal/controllers/gateway/gateway_controller.go:
```diff
@@ -0,0 +1,594 @@
+
+	// GVK information for helper functions
+	group                             = gatewayv1alpha2.Group(gatewayv1alpha2.GroupName)
+	kind                              = gatewayv1alpha2.Kind("HTTPRoute")
```
`kind` in particular can be a constant but the others can't.

To make this implementation more clear I've moved things around considerably, take a look 3502fe7f4b8ec2703238ff3e14c09bb644be552a

shaneutt commented:
> Initial review is that between this being a completely new (and particularly unique) controller and my limited understanding of GWAPI details I probably want a walkthrough of this on a call to review this effectively. Following are some initial comments that are 80% bikeshedding (though I do think the "provisioned" one is important) and 20% primer questions for a walkthrough.
 
> Do we need to get rid of nakedreturn linting? Personal preference but I don't find them particularly readable, and it seems like if anything we should be exempting the individual instances if we really do want it somewhere (though I don't think there's ever any case where it's difficult to explicitly return).

It is to my great dismay that I suppose I'll just have to concede on this point since it's been brought up twice now. Changed in 0fbe4db1ce844331daf762d908784a36c2463266.

> gateway_controller.go differs considerably from the other controllers, which we outline in `hack/generators/controllers/networking/main.go` and template to create `internal/controllers/configuration/zz_generated_controllers.go`. This looks probably necessary, though it'd probably help to cover how and why it differs, especially for stuff that exists elsewhere in the codebase in a different form (class filtering and publish service information).

The main reasons why this differs so greatly from other controllers are:

1. it does _not_ cause any updates which would be processed by the Kong Admin API (unlike every other type currently) and is meant almost entirely for constraints and function of the linked types `HTTPRoute`, `TLSRoute`, `TCPRoute`, `UDPRoute`, e.t.c.
2. it needs to watch multiple types in order to ensure eventual consistency
 
> Is "Pre-Provisioned" and "Provisioned" official nomenclature (the annotations are ours and I didn't see mention of it in a quick search of the GWAPI docs, so guessing no)? My reading of the code suggests that the latter means the Gateway is not actually, well, provisioned, but rather something that needs to be provisioned, whereas my reading of the actual words is that they mean the same thing. Not immediately sure what would be better offhand--maybe "pre-provisioned" versus "controller-provisioned" or "externally-managed" and "controller-managed" to clarify the actor? Would maybe be easier to say knowing roughly how the controller would handle a gateway that isn't already provisioned in the future.

The official language is not yet determined, and in fact I'm assigned to the upstream issue which is meant to determine that: https://github.com/kubernetes-sigs/gateway-api/issues/892 until there's consensus "pre-provisioned" seemed like something we could easily key into later and replace with whatever the official language ends up being.
 
> More naming nits: `obj` makes me think of generic K8S objects (when you actually don't know the GVK) and confused me in the longer functions; I had to stop a few times and scroll back to the signature to realize that it was the Gateway being reconciled or whatever. `gatewayObject` or just `gateway`/`gatewayClass`/etc. as appropriate?

Sure thing, I went with `gateway` and `gatewayClass` and this is now changed in b68a280e46a1c4a36b0adf06a80b88aec742f818

> This seems to have parallel implementations of (or at least related to) existing watch namespace and publish service functionality. How/why do GWAPIs need to handle these differently?

This constraint was the best combination I found of "simple" and "strict" effectively allowing no wiggle room for the initial implementation, and I intentionally wanted to lock things down for the first iteration and PR because it's simple to loosen things later, but the opposite is not as true.

shaneutt commented on internal/controllers/gateway/gateway_controller.go:
```diff
@@ -0,0 +1,610 @@
+		r.Log.Error(err, "failed to list gateways for gatewayclass in watch", "gatewayclass", gwc.GetName())
+		return
+	}
+	for _, gateway := range gateways.Items {
+		if string(gateway.Spec.GatewayClassName) == gwc.GetName() {
+			recs = append(recs, reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Namespace: gateway.Namespace,
+					Name:      gateway.Name,
+				},
+			})
+		}
+	}
```
Move into a function

shaneutt commented:
Some new requirements and concerns came up in a Zoom meeting today which are going to result in a significant rewrite of parts of this PR. As such I'm closing this PR in favor of a new one so we can start with a fresh review as the follow up will be significantly different.
